### ISSUE TITLE
TEC fitting convergence causes image ripples

### ISSUE THREAD

**Issue by rvweeren** [lofar-astron/factor#87]
I have been trying hunting down the stripe and ripple patterns I get on 3C295 (similar stripes are encountered by Wendy and David for 3C295). First I had a look at the visibilities after factor solved and subtracted (peeled) 3C295. (first ~1 hrs of data)

<img width="497" alt="tecfitflaw" src="https://cloud.githubusercontent.com/assets/3879473/15195760/475aaa2c-1797-11e6-9122-ca073d4fead4.png">

Around ~8 pm suddenly the amplitudes have a huge jump up to 80 Jy. (that is how bright 3C295 is). So clearly 3C295 is not properly subtracted at that point. Almost all the bad visibilities come from RS407, but also a few from RS210/508/509. The pattern changes, few hrs later RS407 is normal, but now there is another big jump for quite a while due to RS509, followed by RS210. The patterns seem to depend on the different TEC groups. 

Some more examples at later times (see the time axis, I plot chunks as the data is "chunked" by factor) for RS509/210.

<img width="488" alt="tecfitflaw3" src="https://cloud.githubusercontent.com/assets/3879473/15196084/c1d46fb2-1798-11e6-9f8c-634737c72c8f.png">
<img width="500" alt="tecfitflaw2" src="https://cloud.githubusercontent.com/assets/3879473/15196085/c1d6118c-1798-11e6-8e30-9900f6e12dba.png">

Then I had a look at the actual TEC solutions (not the phases, the actual TEC values, referenced to CS001HBA). At the same times when the bad data occurs there are these TEC jumps. At these times 3C295 is very badly subtracted from the data leaving the 80 Jy residuals and very likely causing the nasty image stripes.  Most problems occur in the first "TEC group" (out of 4 groups), probably because is has most trouble converging due to the stronger phase gradient at low-freqs. 

![testplot_tec](https://cloud.githubusercontent.com/assets/3879473/15197415/55d7ddec-179f-11e6-82f8-3e4af41fd3ca.png)

The problem I am seeing must occur for fainter sources as well, but it will often remain hidden as the ripples will be fainter. (also because the slow Gains timescales are long compared tec fitting timescales and then if say 10% of the data in a solution interval has a jump the actual slow gains might still look ok in the solutions).

Also note that the wrong TEC might be ok for some channels across a TEC group, but the phase gradient across that blocks is wrong.  For example, in my case for RS509 the phase gradient switches sign(!) at these jumps. (so inside a TEC group I get a completely wrong phase gradient screwing up the subtract)

Fist step is I think add TEC plots to the plotting script to fully expose the problem (so plotting actual TEC, not phase, and also not adding the ScalarPhase). We then at least have warning signs. 

The 80 Jy of bad residual data shows how severe this problem. I really hope the new TEC fitting is ready for testing soon and it solves this convergence issue.

**Comment by rvweeren** [comment]
BTW This is my residual (peeled) 3C295 facet image (full bandwidth) showing the striping problem.

<img width="606" alt="afterpeel" src="https://cloud.githubusercontent.com/assets/3879473/15197863/cfc46358-17a1-11e6-88cd-4f3c67bde71c.png">

**Comment by twshimwell** [comment]
This is pretty nasty. I'm not sure how long until the new TEC fitting is ready for testing. In the meantime, if this is a limiting factor, could we attempt to automatically identify these bad time periods for the different antennas (maybe losoto can do this) assuming there are few of them (which I thought normally seems to be the case on HBA data) and then just flagging the outlier data. Not sure how this would build up over say 20 facets, hopefully it wouldn't end up flagging too much or everything.

**Comment by darafferty** [comment]
Instead of flagging, maybe we could replace the bad values with interpolated ones before applying them?

I'm adding the TEC plots now. By the way, I merged the reimage branch into the master one, so I will start adding the new stuff to the master branch again instead of the reimage one.

**Comment by twshimwell** [comment]
Yes that sounds like a better suggestion to me. I wonder if amp v time v freq waterfall plots are also useful as this issue really jumped out in that amp v time plots. I guess it is an exceptionally bright source though.

**Comment by rvweeren** [comment]
It is not as as simple as replacing the TEC. The scalar phase is wrong as well, if you do not adjust the scalar phase things will probably not be any better.

> Not sure how this would build up over say 20 facets, hopefully it wouldn't end up flagging too much or everything.

I think that after 20 facets you have flagged the majority of the distant RS stations (these jumps do not always occur at the same timeslots for different facets). 

I have a solution in mind that would likely fix this issue for at least the 10-15 brighter sources in the field. However, if we have better TEC fitting soon it is not worth working on this. Problem is that the timescales for the TEC fitting implementation/testing are unclear to me. Are we talking about one/two months ? Or more ? Maybe Tammo Jan could respond….

**Comment by darafferty** [comment]
I've now converted all the solve steps to DPPP GainCal, including the TEC+scalarphase ones. However, since the TEC solver is still in development, the fast-phase solve still uses BBS by default. Once the new solver is ready, it can be enabled by setting use_gaincal = True under [calibration]. (I've tested that it at least runs with Tammo Jan's development version of DPPP -- the results were actually quite good even at this stage.)

**Comment by darafferty** [comment]
I've now implemented GainCal TEC+scalarphase solving in frequency blocks, so GainCal will now respect the TEC_block_MHz setting (previously it would only fit a single solution over the whole bandwidth, leading to fairly large errors at the frequency extremes). This functionality is only available on the "tecsolve" branch for now until it's been tested a bit more.  (Note: use_gaincal = True is no longer needed as BBS has been completely removed.)

So far, I find that the new solver performs quite well -- I think that the discretization issue we saw at the last busy week is not really a problem, as it seems to result from the way the solver does the brute-force search (which is done in discrete steps) and then refines the solutions. During the refining stage, the solver will often adjust the scalarphase value rather than the TEC value to get the best fit. In these cases, the TEC solutions remain at the discrete values they had from the brute-force search. But the discretization should not appear in the final sum of TEC+scalarphase (which is all we really care about I guess). More testing is needed to confirm this, though.
